pkg/covermerger/covermerger.go

```diff
@@ -182,32 +173,89 @@ func MergeCSVData(config *Config, reader io.Reader) (map[string]*MergeResult, er
 	return mergeResult, nil
 }
 
-func mergeChanData(c *Config, recordsChan <-chan FileRecord) (map[string]*MergeResult, error) {
+type FileRecords struct {
+	fileName string
+	records  []FileRecord
+}
+
+func mergeChanData(c *Config, recordChan <-chan FileRecord) (map[string]*MergeResult, error) {
+	processedFiles := map[string]int{}
+	frecordChan := make(chan FileRecords)
+	exitCh := make(chan struct{})
+	go func() {
+		targetFile := ""
+		var records []FileRecord
+		for record := range recordChan {
+			select {
+			case <-exitCh:
+				return
+			default:
+			}
+			curTargetFile := record[KeyFilePath]
+			if targetFile == "" {
+				targetFile = curTargetFile
+			}
+			if curTargetFile != targetFile {
+				processedFiles[targetFile]++
+				frecordChan <- FileRecords{
+					fileName: targetFile,
+					records:  records,
+				}
+				records = nil
+				targetFile = curTargetFile
+			}
+			records = append(records, record)
+		}
+		processedFiles[targetFile]++
+		frecordChan <- FileRecords{
+			fileName: targetFile,
+			records:  records,
+		}
+		close(frecordChan)
+	}()
 	stat := make(map[string]*MergeResult)
-	targetFile := ""
-	var records []FileRecord
-	processedFiles := map[string]struct{}{}
-	for record := range recordsChan {
-		curTargetFile := record[KeyFilePath]
-		if targetFile == "" {
-			targetFile = curTargetFile
+	waitCh := make(chan struct{}, 1)
+	errCh := make(chan error, 1)
+	go func() {
```

Comment:
Do you really need to run the code below in a separate goroutine? The code does not block on anything + you won't need the `waitCh` channel.

Reply:
Thanks for the https://pkg.go.dev/golang.org/x/sync/errgroup proposal.

Reply:
done
looks much better